org.eclipse.xtext.xbase/src/org/eclipse/xtext/xbase/typesystem/internal/ExpressionAwareStackedResolvedTypes.java

```diff
@@ -24,6 +29,13 @@ protected ExpressionAwareStackedResolvedTypes(ResolvedTypes parent, XExpression
 		super(parent);
 		this.expression = expression;
 	}
+	
+	/**
+	 * @since 2.30
+	 */
+	protected XExpression expression() {
```

Comment:
by which criteria do we decide that in StackedResolvedTypes subclasses in protected or public?

Reply:
Good question. Gut feeling I'd say. At least these things should be accessible for override, even though all this is supposed to be internal and not API for the average use-case.

If we want to expose this as APIs, I'd make that a separate ticket and consolidated effort.